### PR TITLE
Cleanup project system code

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Commit at-a-time review recommended: as I was investigating crash reports last week I did a little cleanup so I could actually understand what the code was doing. I also fixed up tests so they actually were testing the product correctly. There is no functional change here.

Since this is really mostly a test-changes-only my gut is the ask mode template doesn't apply here. (If it does, most of the questions don't make sense.)

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
And @Pilchie or @jinujoseph, as code reviews happen and CI integrates continuously, does this even need an ask mode template?

**Comment by jasonmalinowski** [comment]
I should clarify: I wouldn't want this in master, as further perf work improvements in dev15.7.x would benefit from this.

File: src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs [hunk 1]
```diff
@@ -108,34 +108,65 @@ public void AddCyclicProjectReferencesDeep()
...
+            using (var environment = new TestEnvironment())
+            {
+                var project1 = CreateCSharpProject(environment, "project1");
+                project1.SetBinOutputPathAndRelatedData(@"c:\project1.dll");
+
+                var project2 = CreateCSharpProject(environment, "project2");
+                project2.SetBinOutputPathAndRelatedData(@"c:\project2.dll");
+
+                // since this is known to be the output path of project1, the metadata reference is converted to a project reference
+                project2.OnImportAdded(@"c:\project1.dll", "project1");
```
Comment (heejaechang) [inline src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs]: so we no longer treat p2p reference and metadata reference differently? any reference that points to same path to project becomes p2p reference?

Comment (heejaechang) [reply src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs]: what happen if order is revered, reference is added first, and then project is added? will existing reference turns to p2p reference?

File: src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs [hunk 2]
```diff
@@ -108,34 +108,65 @@ public void AddCyclicProjectReferencesDeep()
...
+
+                Assert.Single(project2.GetCurrentProjectReferences().Where(pr => pr.ProjectId == project1.Id));
+
+                project1.Disconnect();
+
+                // Verify project reference updated after bin path change.
+                Assert.Empty(project2.GetCurrentProjectReferences());
+                Assert.Single(project2.GetCurrentMetadataReferences().Where(r => r.FilePath == @"c:\project1.dll"));
+
+                project2.Disconnect();
```
Comment (heejaechang) [inline src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs]: here we probably want to create project1 again and add it back? and see whether projec2 gets that as p2p reference?

File: src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs]: @heejaechang Yes, we've been doing that for awhile now. This is just fixing the tests to actually test that.

And yes, either order works.

File: src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs]: I think that's already covered in `AddingReferenceToProjectMetadataPromotesToProjectReference` and `AddingProjectReferenceAndUpdateReferenceBinPath` tests effectively?

Comment (heejaechang) [reply src/VisualStudio/CSharp/Test/ProjectSystemShim/LegacyProject/CSharpReferencesTests.cs]: I see. good then.

**Comment by jinujoseph** [comment]
@jasonmalinowski to be clear you want to merge this only after [link](https://devdiv.visualstudio.com/DevDiv/_queries/edit/556821/?triage=false) is fixed right ?

**Comment by jinujoseph** [comment]
ah nevermind , this is targeted for 15.7 , we are good

**Comment by jasonmalinowski** [comment]
@jinujoseph Doesn't matter, we could put this into 15.6. From the unit test perspective this is touching, our code already worked fine.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject.cs [hunk 3]
```diff
@@ -169,94 +169,5 @@ private static bool GetIsWebsiteProject(IVsHierarchy hierarchy)
...
-            var delta = set.Count - noReferenceOutputAssemblies.Count - (projectReferences.Length + metadataReferences.Length);
-            if (delta == 0)
-            {
-                return;
-            }
-
-            // okay, two has different set of dlls referenced. check special Microsoft.VisualBasic case.
-            if (delta != 1)
-            {
-                //// Contract.Requires(false, "different set of references!!!");
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject.cs]: lol... debug only code... with commented out asserts...

**Pull request merged**